inmem.go

```diff
@@ -0,0 +1,244 @@
+package clover
+
+import (
+	"sort"
+	"sync"
+)
+
+type collection map[string]*Document
+
+type inMemEngine struct {
+	sync.RWMutex
+	collections map[string]collection
+}
+
+func newInMemoryStoreEngine() StorageEngine {
+	return &inMemEngine{
+		RWMutex:     sync.RWMutex{},
+		collections: make(map[string]collection),
+	}
+}
+
+func InMemoryDB() (*DB, error) {
+	return &DB{
+		dir:    "",
+		engine: newInMemoryStoreEngine(),
+	}, nil
+}
+
+// Close implements StorageEngine
+func (*inMemEngine) Close() error {
+	return nil
+}
+
+// CreateCollection implements StorageEngine
+func (e *inMemEngine) CreateCollection(name string) error {
+	e.Lock()
+	defer e.Unlock()
+	if ok, _ := e.HasCollection(name); ok {
+		return ErrCollectionExist
+	}
+	e.collections[name] = make(collection)
+	return nil
+}
+
+// Delete implements StorageEngine
+func (e *inMemEngine) Delete(q *Query) error {
+	e.Lock()
+	defer e.Unlock()
+
+	c, ok := e.collections[q.collection]
+	if !ok {
+		return ErrCollectionNotExist
+	}
+
+	for key, d := range c {
+		if q.satisfy(d) {
+			delete(c, key)
+		}
+	}
+
+	return nil
+}
+
+// DeleteById implements StorageEngine
+func (e *inMemEngine) DeleteById(collectionName string, id string) error {
+	e.Lock()
+	defer e.Unlock()
+
+	c, ok := e.collections[collectionName]
+	if !ok {
+		return ErrCollectionNotExist
+	}
+
+	delete(c, id)
+
+	return nil
+}
+
+func (e *inMemEngine) hasCollection(collectionName string) bool {
+	_, ok := e.collections[collectionName]
+	return ok
+}
+
+// DropCollection implements StorageEngine
+func (e *inMemEngine) DropCollection(name string) error {
+	e.Lock()
+	defer e.Unlock()
+
+	if ok, _ := e.HasCollection(name); ok {
```

Comment:
Use the private version `hasCollection` here (HasCollection() will call e.RLock(), so you could get a deadlock).